### PR TITLE
Create DevSession for integration tests

### LINKED ISSUE #5549: Write utility function that acts as `odo push` but uses `odo dev`
Write utility function that acts as odo push but uses odo dev

this function will be used in e2e and integrations tests

- [ ] write the function 
- [ ] replace `odo push` with this function in at least one test file

/kind task
/area testing

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind tests

**What does this PR do / why we need it:**

This PR abstracts running `odo dev` for integration tests

**Which issue(s) this PR fixes:**

Fixes #5549 

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: tests/helper/helper_dev.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	# Starting a session for a series of tests and stopping the session after the tests:
+
+	```
+	When("running dev session", func() {
+		var devSession DevSession
+		BeforeEach(func() {
+			devSession = helper.StartDevMode()
+		})
+		AfterEach(func() {
+			session.Stop()
```
Comment (anandrkskd) [inline tests/helper/helper_dev.go:20]: It should  be `devSession` right?

Comment (feloy) [reply tests/helper/helper_dev.go]: Yes right, thanks

**Comment by anandrkskd** [comment]
@feloy can you also help understand in what tests/scenarios, where `Rundevmode` needs to be used and where  `startdevmode` needs to be used? 
I guess that would help maintain tests understandable.

**Review by rm3l — COMMENTED** [review]
LGTM from a code perspective 👍🏿. The new helper functions look really nice and easy to work with.

Do you know why this test is reportedly not passing? I guess it is unrelated to the changes here, but wanted to be sure :)

```
odo devfile push command tests
/go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:18
  when Create and push java-springboot component
  /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:687
    should execute default build and run commands correctly [It]
    /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:693
    Unexpected error:
              
        <*errors.errorString | 0xc0002e2040>: {
            s: "cmd [ps -ef] failed with error command terminated with exit code 1\n on pod tuvmph-app-7876b9c979-t9n89",
        }
        cmd [ps -ef] failed with error command terminated with exit code 1
         on pod tuvmph-app-7876b9c979-t9n89
    occurred
    /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:710
```

**Comment by feloy** [comment]
> ```
> odo devfile push command tests
> /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:18
>   when Create and push java-springboot component
>   /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:687
>     should execute default build and run commands correctly [It]
>     /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:693
>     Unexpected error:
>               
>         <*errors.errorString | 0xc0002e2040>: {
>             s: "cmd [ps -ef] failed with error command terminated with exit code 1\n on pod tuvmph-app-7876b9c979-t9n89",
>         }
>         cmd [ps -ef] failed with error command terminated with exit code 1
>          on pod tuvmph-app-7876b9c979-t9n89
>     occurred
>     /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:710
> ```

This error happens regularly on tests, but Icannot find the reason of this error. @anandrkskd @kadel any idea?

**Comment by feloy** [comment]
/test v4.9-integration-e2e

```
<*errors.errorString | 0xc0002e2040>: {
            s: "cmd [ps -ef] failed with error command terminated with exit code 1\n on pod tuvmph-app-7876b9c979-t9n89",
        }
        cmd [ps -ef] failed with error command terminated with exit code 1
         on pod tuvmph-app-7876b9c979-t9n89
    occurred
    /go/src/github.com/redhat-developer/odo/tests/integration/devfile/cmd_devfile_push_test.go:710
```

**Comment by anandrkskd** [comment]
> This error happens regularly on tests, but Icannot find the reason of this error. @anandrkskd @kadel any idea?

This is the flake that we are experiencing on PRs and periodic jobs and are tracked under https://github.com/redhat-developer/odo/issues/5506, this flake used to happen before, but it disappeared during v2 development.

I think we should make this one priority for our tests, try to resolve this ASAP.

**Comment by feloy** [comment]
/test v4.9-integration-e2e

infra error

**Review by rm3l — APPROVED** [review]

**Comment by feloy** [comment]
/test v4.9-integration-e2e

```
dial tcp: lookup api.ci-op-fgdj1r5w-f22a6.origin-ci-int-aws.dev.rhcloud.com on 172.30.0.10:53: read udp 10.131.196.96:56164->172.30.0.10:53: i/o timeout.
```

File: tests/integration/devfile/cmd_dev_test.go [hunk 2]
```diff
@@ -52,30 +52,25 @@ var _ = Describe("odo dev command tests", func() {
 			Expect(helper.VerifyFileExists(".odo/env/env.yaml")).To(BeFalse())
 		})
 		It("should show validation errors if the devfile is incorrect", func() {
-			session := helper.CmdRunner("odo", "dev")
-			defer session.Kill()
-			helper.WaitForOutputToContain("Waiting for something to change", 180, 10, session)
-			helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "kind: run", "kind: build")
-			helper.WaitForOutputToContain(watch.PushErrorString, 180, 10, session)
+			helper.RunDevMode(func(session *gexec.Session) {
```
[the remark this answers is not in the document]
Comment (anandrkskd) [reply tests/integration/devfile/cmd_dev_test.go:55]: I think, we are passing the function (L55-L57) and it's executed [here](https://github.com/redhat-developer/odo/pull/5569/files#diff-61062e4f376778f1c4203c233f7357bd31ded114bf3e09248fffbf83bb8e0247R113) after the session is defined [here](https://github.com/redhat-developer/odo/pull/5569/files#diff-61062e4f376778f1c4203c233f7357bd31ded114bf3e09248fffbf83bb8e0247R110).

[the remark this answers is not in the document]
Comment (feloy) [reply tests/integration/devfile/cmd_dev_test.go:55]: It is the same "handler" or "callback" pattern. This `RunDevMode` method will run the Dev command and will call the function you are passing as argument when the Dev mode is started. The session is passed as an arg to this inline function from the `runDevMode` one.

**Comment by feloy** [comment]
/test v4.9-integration-e2e

infra error

**Review by rm3l — APPROVED** [review]

**Comment by feloy** [comment]
/test v4.10-integration-e2e

infra error?

**Comment by dharmit** [comment]
> /test v4.10-integration-e2e
> 
> infra error?

```
the pod ci-op-5p7x347d/integration-e2e-integration-e2e-steps was deleted without completing after 28m29s (failed containers: )
```
IMO, yes. Hitting pod deletion errors on https://github.com/redhat-developer/odo/pull/5526#issuecomment-1075030523.

**Pull request merged**